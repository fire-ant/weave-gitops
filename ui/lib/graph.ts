// A collection of helper functions to render a graph of kubernetes objects
// with in the context of their parent-child relationships.
import _ from "lodash";
import { convertResponse } from "../hooks/objects";
import { Core } from "./api/core/core.pb";
import { GroupVersionKind, Kind } from "./api/core/types.pb";
import { FluxObject } from "./objects";

// Kubernetes does not allow us to query children by parents.
// We keep a list of common parent-child relationships
// to look up children recursively.
export const PARENT_CHILD_LOOKUP = {
  Deployment: {
    group: "apps",
    version: "v1",
    kind: "Deployment",
    children: [
      {
        group: "apps",
        version: "v1",
        kind: "ReplicaSet",
        children: [{ version: "v1", kind: "Pod" }],
      },
    ],
  },
  StatefulSet: {
    group: "apps",
    version: "v1",
    kind: "StatefulSet",
    children: [{ version: "v1", kind: "Pod" }],
  },
};

const objID = (obj: FluxObject) =>
  `${obj.clusterName}/${obj.namespace}/${obj.type}/${obj.name}`;

const sortFn = (a, b) => {
  return objID(a).localeCompare(objID(b));
};

export const getChildrenRecursive = async (
  client: typeof Core,
  namespace: string,

  object: FluxObject,
  clusterName: string,
  lookup: any
) => {
  const children: FluxObject[] = [];

  const k = lookup[object.type];

  if (k && k.children) {
    for (let i = 0; i < k.children.length; i++) {
      const child: GroupVersionKind = k.children[i];

      const res = await client.GetChildObjects({
        parentUid: object.uid,
        namespace,
        groupVersionKind: child,
        clusterName: clusterName,
      });

      for (let q = 0; q < res.objects.length; q++) {
        const c = convertResponse(null, res.objects[q]);
        // Dive down one level and update the lookup accordingly.
        await getChildrenRecursive(client, namespace, c, clusterName, {
          [child.kind]: child,
        });
        children.push(c);
      }
    }
  }
  // Mutates
  children.sort(sortFn);
  object.children = children;
};

// Gets the "child" objects that result from an Application
export const getChildren = async (
  client: typeof Core,
  automationName,
  namespace,
  automationKind: Kind,
  kinds: GroupVersionKind[],
  clusterName
): Promise<FluxObject[]> => {
  const { objects } = await client.GetReconciledObjects({
    automationName,
    namespace,
    automationKind,
    kinds,
    clusterName,
  });

  const fluxObjs = _.map(objects, (o) => convertResponse(null, o));
  fluxObjs.sort(sortFn);

  const result: FluxObject[] = [];
  for (let o = 0; o < fluxObjs.length; o++) {
    const obj = fluxObjs[o];
    await getChildrenRecursive(
      client,
      namespace,
      obj,
      clusterName,
      PARENT_CHILD_LOOKUP
    );
    result.push(obj);
  }
  const flat = _.flatten(result);

  return flat;
};
