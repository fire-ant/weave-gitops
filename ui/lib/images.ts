/* eslint-disable */
// @ts-nocheck

import bg from "data-url:../images/bg-circles.png";
import failedSrc from "data-url:../images/failed.svg";
import fluxIconHoverSrc from "data-url:../images/flux-icon-hover.svg";
import fluxIconSrc from "data-url:../images/flux-icon.svg";
import logotype from "data-url:../images/logotype.svg";
import reconcileSrc from "data-url:../images/reconcile.svg";
import signInWheel from "data-url:../images/SignInWheel.svg";
import successSrc from "data-url:../images/success.svg";
import suspendedSrc from "data-url:../images/suspended.svg";
import weaveG from "data-url:../images/weaveG.svg";
import weaveLogo from "data-url:../images/WeaveLogo.svg";

export default {
  bg,
  failedSrc,
  weaveG,
  logotype,
  signInWheel,
  successSrc,
  suspendedSrc,
  weaveLogo,
  reconcileSrc,
  fluxIconSrc,
  fluxIconHoverSrc,
};
