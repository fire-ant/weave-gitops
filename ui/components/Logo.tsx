import * as React from "react";
import { Link } from "react-router-dom";
import styled from "styled-components";
import images from "../lib/images";
import { V2Routes } from "../lib/types";
import { Fade } from "../lib/utils";
import Flex from "./Flex";

type Props = {
  className?: string;
  collapsed: boolean;
  link?: string;
};

function Logo({ className, link, collapsed }: Props) {
  return (
    <Flex className={className} wide>
      <Link to={link || V2Routes.Automations}>
        <img src={images.weaveG} />
      </Link>
      <Fade fade={collapsed}>
        <Link to={link || V2Routes.Automations}>
          <img src={images.logotype} />
        </Link>
      </Fade>
    </Flex>
  );
}

export default styled(Logo)`
  a {
    display: flex;
    align-items: center;
  }
  img:first-child {
    margin-right: ${(props) => props.theme.spacing.xs};
  }
  img {
    width: auto;
    height: 32px;
  }
  padding-left: 20px;
  //nav width: 200px - space btwn nav and content: 24px - content padding: 24px. All together 248px - 20 for left padding to line up with detail page titles.
  width: ${(props) => (props.collapsed ? "92px" : "228px")};
  transition: width 0.5s;
`;
