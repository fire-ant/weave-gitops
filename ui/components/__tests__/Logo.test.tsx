import "jest-styled-components";
import React from "react";
import renderer from "react-test-renderer";
import { withContext, withTheme } from "../../lib/test-utils";
import Logo from "../Logo";

describe("Logo", () => {
  describe("snapshots", () => {
    it("renders open view", () => {
      const tree = renderer
        .create(withTheme(withContext(<Logo collapsed={false} />, "", {})))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
    it("renders collapsed view", () => {
      const tree = renderer
        .create(withTheme(withContext(<Logo collapsed={true} />, "", {})))
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
