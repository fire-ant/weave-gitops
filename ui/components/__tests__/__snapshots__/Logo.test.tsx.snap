// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logo snapshots renders collapsed view 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: 100%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c3 {
  opacity: 0;
  -webkit-transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  pointer-events: none;
}

.c1 {
  padding-left: 20px;
  width: 92px;
  -webkit-transition: width 0.5s;
  transition: width 0.5s;
}

.c1 a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 img:first-child {
  margin-right: 8px;
}

.c1 img {
  width: auto;
  height: 32px;
}

<div
  className="c0 c1"
>
  <a
    href="/applications"
    onClick={[Function]}
  >
    <img
      src=""
    />
  </a>
  <div
    className="c2 c3"
  >
    <a
      href="/applications"
      onClick={[Function]}
    >
      <img
        src=""
      />
    </a>
  </div>
</div>
`;

exports[`Logo snapshots renders open view 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
  width: 100%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c3 {
  opacity: 1;
  -webkit-transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c1 {
  padding-left: 20px;
  width: 228px;
  -webkit-transition: width 0.5s;
  transition: width 0.5s;
}

.c1 a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 img:first-child {
  margin-right: 8px;
}

.c1 img {
  width: auto;
  height: 32px;
}

<div
  className="c0 c1"
>
  <a
    href="/applications"
    onClick={[Function]}
  >
    <img
      src=""
    />
  </a>
  <div
    className="c2 c3"
  >
    <a
      href="/applications"
      onClick={[Function]}
    >
      <img
        src=""
      />
    </a>
  </div>
</div>
`;
